auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java

```diff
@@ -61,38 +65,64 @@ public void startSignIn(@NonNull final String email,
                     .build();
         }
 
-        // Kick off the flow including signing in, linking accounts, and saving with SmartLock
-        getAuth().signInWithEmailAndPassword(email, password)
-                .continueWithTask(new Continuation<AuthResult, Task<AuthResult>>() {
-                    @Override
-                    public Task<AuthResult> then(@NonNull Task<AuthResult> task) throws Exception {
-                        // Forward task failure by asking for result
-                        AuthResult result = task.getResult(Exception.class);
+        // We first check if we are trying to upgrade an anonymous user
+        // Calling linkWithCredential will fail, so we can save an RPC call
+        // We just need to validate the credential
+        if (AnonymousUpgradeUtils.canUpgradeAnonymous(getAuth(), getArguments())) {
+            final AuthCredential credToValidate = EmailAuthProvider.getCredential(email, password);
 
-                        // Task succeeded, link user if necessary
-                        if (credential == null) {
-                            return Tasks.forResult(result);
-                        } else {
-                            return result.getUser()
-                                    .linkWithCredential(credential)
-                                    .continueWithTask(new ProfileMerger(outputResponse))
-                                    .addOnFailureListener(new TaskFailureLogger(TAG, "linkWithCredential+merge failed."));
+            AnonymousUpgradeUtils.validateCredential(FirebaseApp.getInstance(), credToValidate)
+                    .continueWith(new Continuation<AuthResult, Void>() {
```

Comment:
Seems like there's no need for a `Continuation` here since the chain ends.  Maybe it should be an `OnCompleteListener`?

Reply:
Done.